Desktop Access without Active Directory (#19311)
gravitational/teleport: docs/pages/desktop-access/getting-started.mdx

---
title: Getting Started with Desktop Access
description: Access Windows systems through Teleport Desktop Access
---

> 

Reviewer: How about a note at the top to help orient the reader and ensure they're on the right page? Something like:

```
Using Active Directory?

This guide sets up Teleport Desktop Access for standalone hosts that are not
joined to an Active Directory domain. To integrate Teleport with AD, see our
Active Directory guide.
```

Author: Yes, I still need to write the intro to the doc. This is a draft PR, staged as such while I make sure the technical details are correct as they are confirmed.

## Prerequisites

- An existing Teleport cluster with one of the following versions:

>   **Open Source or Enterprise:** version 8.0 or newer

Reviewer: For non-AD this will be 12.0 or newer.

Author: Yup, still need to update a lot in this. Draft PR.

Import the certificate to your Windows certificate trust. Right click on the
certificate and select **Install Certificate**:

![Install the Teleport root Certificate](../../img/desktop-access/install-root-certificate@2x.png)

From the Certificate Import Wizard under **Store Location** select **Local Machine**:

![Select Local Machine](../../img/desktop-access/certificate-import-step1@2x.png)

At the next screen, select **Place all certificates in the following store**.
Click **Browse** and select "Trusted Root Certification Authorities":

![Trusted Root Certification Authorities](../../img/desktop-access/trusted-root-certification-authorities@2x.png)

At the next screen, click **Finish** to complete the certificate import.

The Teleport DLL file provides support for Teleport's certificate-based
authentication for Windows systems not connected to an Active Directory Domain.

[Download the Teleport DLL](#LinkToCome), and register it in your system. From a Powershell
window opened with Administrator Privleges, run:

```code
$ cp teleport.dll C:\Windows\System32\teleport.dll
$ regsvr32.exe C:\Windows\System32\teleport.dll
```

Reboot the Windows system after installing the DLL.

> ### Disable network level authenticate for RDP

Reviewer: ```suggestion
### Disable Network Level Authentication for RDP
```

---

> title: Getting Started with Desktop Access

Author: ```suggestion
title: Getting Started with Windows Access
```

Reviewer: Desktop Access is fine. Let's keep it.

> description: Access Windows systems through Teleport Desktop Access

Author: ```suggestion
description: Access Windows systems through Teleport using passwordless authentication.
```

---

This guide will help you configure Teleport to provide secure, passwordless access
to Windows desktops. This configuration does not require an Active Directory domain.

If you are running Windows in an AD environment, consider integrating Teleport
with Active Directory for automatic discovery by reading [Desktop Access with

> Active Directory](./active-directory.mdx). You can also mix Active Directory

Reviewer 2: ```suggestion
Active Directory](./active-directory.mdx). You can also mix the Teleport Active Directory
```

integration with the static host definitions described below.

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="12.0"
>
Windows Access without Active Directory is available starting from Teleport `v12`.
Previous versions of Teleport can implement Windows Access by integrating with
an [Active Directory domain](./active-directory.mdx).
</Details>

<Admonition type="warning">
Teleport Desktop Access is not yet compatible with Windows Server 2022.
</Admonition>

## Prerequisites

> - (!docs/pages/includes/edition-prereqs-tabs.mdx!)

Reviewer 2: Not to bring this issue up again, but I'm not sure what the outer bullet indicates here. What category do the inner bullets belong to that unifies them under the outer bullet? 

If we do want to add the bullets within the `Tabs` component under the same outer bullet, maybe we could include something like:

```markdown
- Make sure you have a running Teleport cluster and client tools:

  (!docs/pages/includes/edition-prereqs-tabs.mdx!)
```

Author: Adjusted, though we should definitely go back and review this partial's format for better inclusion in this list.

- A Linux server to run the Teleport Desktop Access service on.
  You can reuse an existing server running any other Teleport instance.

> - (!docs/pages/includes/tctl.mdx!)

Reviewer 2: Putting a `Details` box in a bullet like this looks sort of awkward to me, like we're trying to work against the design of our docs components. Is there an alternative?

> ## Step 1/4. Prepare Windows

Reviewer 2: I think this section would benefit from a short paragraph that explains the architecture of the setup we're working through in the guide. That way, it would be easier for me to form a mental model that I can use to make sense of all the instructions.

Author: Added below.

### Import the Teleport root certificate

Use `tctl` to export the Teleport root certificate:

```code
$ tctl auth export --type=windows > teleport.cer
```

Copy this certificate to your Windows system, if you didn't run `tctl` from there.

### Install the Teleport service for Windows

From the Windows system, download the [Teleport service installer](#LinkToCome) {/*Rename as needed */}
and run it. When prompted, select the Teleport certificate file from the previous step.
Once complete, reboot the system.

## Step 2/4. Install the Teleport Desktop Service

On the Linux host where you will run the Desktop Service, install Teleport (if

> you plan on running the Windows protocol on the same host as an existing Teleport

Reviewer 2: ```suggestion
you plan on running the Desktop Service on the same host as an existing Teleport
```
For precision, since a protocol isn't what you would be running on the host

Author: Yea... this was based on some input I got on naming conventions that have since changed...

```suggestion
you plan on running the Windows Service on the same host as an existing Teleport
```

service, you can skip this step):

> (!docs/pages/includes/install-linux.mdx!)

Reviewer 2: This partial has its own intro paragraph that conflicts with the one in this guide:

<img width="907" alt="image" src="https://user-images.githubusercontent.com/25250668/214699702-e1cb3681-3871-4879-b872-69825feff599.png">

Author: Imported a commit from another PR that makes the partial more compatible with preceding content from the docs it's called in.

> Define the Desktop Access service and list the Windows systems configured for

Reviewer 2: ```suggestion
Enable the Desktop Service and list the Windows systems configured for
```

```

</TabItem>
</Tabs>

## Step 3/4. Configure Windows access

In order to gain access to a remote desktop, a Teleport user needs to have the
appropriate permissions for that desktop.

Create the file `windows-desktop-admins.yaml`:

```yaml
kind: role
version: v6
metadata:
  name: windows-desktop-admins
spec:
  allow:
    windows_desktop_labels:
      "*": "*"
    windows_desktop_logins: ["Administrator", "alice"]
```

You can restrict access to specific hosts by defining values for
`windows_desktop_labels`, and adjust the array of usernames this role has access
to in `windows_desktop_logins`.

<Admonition type="warning" title="RBAC Configuration">
Ensure that each Teleport user is only assigned Windows logins that they should
be allowed to access.
</Admonition>

Apply the new role to your cluster:

```code
$ tctl create -f windows-desktop-admins.yaml
```

(\!docs/pages/includes/add-role-to-user.mdx role="windows-desktop-admins" \!)

See the [RBAC page](./rbac.mdx) for more information about setting up
Windows Desktop Access permissions.

See the [Access Controls Getting Started](../access-controls/getting-started.mdx#step-13-add-local-users-with-preset-roles)

> guide for instructions on how to create or update a user with a given role.

Reviewer 2: I think these should go in a "Next steps" section since they're not required to finish this guide.

Active Directory](./active-directory.mdx). You can also mix Active Directory
integration with the static host definitions described below.

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="12.0"
>
Windows Access without Active Directory is available starting from Teleport `v12`.
Previous versions of Teleport can implement Windows Access by integrating with
an [Active Directory domain](./active-directory.mdx).
</Details>

<Admonition type="warning">
Teleport Desktop Access is not yet compatible with Windows Server 2022.
</Admonition>

## Prerequisites

- (!docs/pages/includes/edition-prereqs-tabs.mdx!)

- A Linux server to run the Teleport Desktop Access service on.
  You can reuse an existing server running any other Teleport instance.

- (!docs/pages/includes/tctl.mdx!)

## Step 1/4. Prepare Windows

### Import the Teleport root certificate

Use `tctl` to export the Teleport root certificate:

```code
$ tctl auth export --type=windows > teleport.cer
```

Copy this certificate to your Windows system, if you didn't run `tctl` from there.

### Install the Teleport service for Windows

> From the Windows system, download the [Teleport service installer](#LinkToCome) {/*Rename as needed */}

Author: The link on this line needs to be replaced with the final name and download location of the installer.

> and run it. When prompted, select the Teleport certificate file from the previous step.

Reviewer: There's also a headless install option that might be worth documenting here.

Author: I'd be happy to document it once I learn about it myself.

Reviewer: You should be able to invoke the same installer from the command line with a `--help` flag to see the options.

Reviewer: @probakowski please advise ☝️

Reviewer 3: It turns out it will show correct output if you run it (with `--help`) from elevated prompt. If you do it from normal prompt it will show UAC screen and then start new elevated console which will close immediately as the program exits (same thing will happen if you create shortcut with `--help` parameter and run it by double click)

Author: I've added a tip below.

Once complete, reboot the system.

## Step 2/4. Install the Teleport Desktop Service

On the Linux host where you will run the Desktop Service, install Teleport (if
you plan on running the Windows protocol on the same host as an existing Teleport
service, you can skip this step):

(!docs/pages/includes/install-linux.mdx!)

Define the Desktop Access service and list the Windows systems configured for
access through Teleport in `/etc/teleport.yaml`.

```yml

> version: v6

Reviewer: ```suggestion
version: v3
```

Roles were bumped to v6, but the latest config version is still v3.

teleport:

>   nodename: example.teleport.com

Reviewer: This is also missing the auth/proxy server address.

Author: Added.

windows_desktop_service:
  enabled: yes

>   listen_addr: 0.0.0.0:3028

Reviewer: We should probably prefer to use tunneling like we do in the other guide. (Connect to proxy on 443/3080)

Author: I removed the key based on the other guide.

>   non_ad_hosts:

Reviewer: Let's add this new field to the config reference as well.

Author: Done.

depend on how you installed Teleport (from a system package or a TAR archive):

<Tabs>
<TabItem label="System Package">

```code
$ sudo systemctl start teleport.service
```

</TabItem>
<TabItem label="Tar Archive">

```code
$ sudo teleport install systemd --output=/etc/systemd/system/teleport.service;
$ sudo systemctl enable teleport;
$ sudo systemctl start teleport;
```

</TabItem>
</Tabs>

## Step 3/4. Configure Windows access

In order to gain access to a remote desktop, a Teleport user needs to have the
appropriate permissions for that desktop.

Create the file `windows-desktop-admins.yaml`:

```yaml
kind: role
version: v6
metadata:
  name: windows-desktop-admins
spec:
  allow:
    windows_desktop_labels:
      "*": "*"
    windows_desktop_logins: ["Administrator", "alice"]
```

You can restrict access to specific hosts by defining values for
`windows_desktop_labels`, and adjust the array of usernames this role has access
to in `windows_desktop_logins`.

<Admonition type="warning" title="RBAC Configuration">
Ensure that each Teleport user is only assigned Windows logins that they should
be allowed to access.

> Usernames shared between domain and local users will create login conflicts.

Reviewer: Since this guide is about non-AD, I don't think domain users are are play. Should we remove this?

---

> title: Getting Started with Windows Access

Reviewer: ```suggestion
title: Getting Started with Desktop Access
```

The pages still appear under a `desktop-access` URL and the menu bar on the left still says "Desktop Access." Let's just stick with desktop access for now.

For open source Teleport, consider integrating Teleport with Active Directory
for automatic discovery by reading [Desktop Access with Active Directory](./active-directory.mdx).

Teleport Enterprise users can also mix the Teleport Active Directory integration
with the static host definitions described below.
</Admonition>

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="12.0"
>
Windows Access without Active Directory is available starting from Teleport `v12`.
Previous versions of Teleport can implement Windows Access by integrating with
an [Active Directory domain](./active-directory.mdx).
</Details>

<Admonition type="warning">
Teleport Desktop Access is not yet compatible with Windows Server 2022.
</Admonition>

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- A Linux server to run the Teleport Desktop Access service on.
  You can reuse an existing server running any other Teleport instance.

(!docs/pages/includes/tctl.mdx!)

## Step 1/4. Prepare Windows

In this section we'll import the Teleport certificate authority (**CA**) file

> to your Windows system, and prepare the it for passwordless access through Teleport.

Reviewer 2: ```suggestion
to your Windows system, and prepare it for passwordless access through Teleport.
```

### Import the Teleport root certificate

> Use `tctl` to export the Teleport root certificate:

Reviewer: ```suggestion
Use `tctl` to export the Teleport user certificate authority:
```

```code
$ tctl auth export --type=windows > teleport.cer
```

Copy this certificate to your Windows system, if you didn't run `tctl` from there.

### Install the Teleport service for Windows

From the Windows system, download the [Teleport service installer](#LinkToCome) {/*Rename as needed */}
and run it. When prompted, select the Teleport certificate file from the previous step.
Once complete, reboot the system.

## Step 2/4. Install the Teleport Desktop Service

On the Linux host where you will run the Desktop Service, install Teleport (if

> you plan on running the Windows Service on the same host as an existing Teleport

Reviewer 2: ```suggestion
you plan on running the Teleport Desktop Service on the same host as an existing Teleport
```
For consistency with the Core Concepts page and other parts of this doc

service, you can skip this step):

(!docs/pages/includes/install-linux.mdx!)

Enable the Desktop Service and list the Windows systems configured for

> access through Teleport in `/etc/teleport.yaml`.

Reviewer: Maybe we should be more explicit that inventory management for non-AD is currently the user's responsibility.

Maybe something like:

```
Without Active Directory, Teleport cannot automatically discover your desktops.
To register desktops with Teleport, you can define them in your configuration file
or use Teleport's API to build your own integration.
```

Author: Updated.

```yml
version: v3
teleport:
  nodename: example.teleport.com
  proxy_server: teleport-proxy.example.com:443

> windows_desktop_service:

Reviewer: `auth_service` and `proxy_service` (and maybe `ssh_service` too, IIRC) are enabled by default, so even though we're trying to keep this config snippet small, we should probably explicitly disable them here.

Author: Added.

  enabled: yes
  non_ad_hosts:
    - 192.0.2.156
```

<Details title="Add labels to hosts">
You can attach labels to your Windows hosts by matching to their hostnames.
For example, to add the `cloud: ec2` label to hosts with EC2 private IP DNS names:

```diff
version: v3
teleport:
  nodename: example.teleport.com
  proxy_server: teleport-proxy.example.com:443
windows_desktop_service:
  enabled: yes
  non_ad_hosts:
-    - 192.0.2.156
+    - ip-192-0-2-156.us-east-2.compute.internal

> + host_labels:

Reviewer: @probakowski can you confirm that `host_labels` apply to `non_ad_hosts` too? (they should, but I'm not sure if we implemented it)

Author: I'm 90% sure I confirmed this through testing.

Reviewer 3: Just confirmed that it does apply

```code
$ sudo teleport install systemd --output=/etc/systemd/system/teleport.service;
$ sudo systemctl enable teleport;
$ sudo systemctl start teleport;
```

</TabItem>
</Tabs>

## Step 3/4. Configure Windows access

In order to gain access to a remote desktop, a Teleport user needs to have the
appropriate permissions for that desktop.

Create the file `windows-desktop-admins.yaml`:

```yaml
kind: role
version: v6
metadata:
  name: windows-desktop-admins
spec:
  allow:
    windows_desktop_labels:
      "*": "*"
    windows_desktop_logins: ["Administrator", "alice"]
```

You can restrict access to specific hosts by defining values for
`windows_desktop_labels`, and adjust the array of usernames this role has access
to in `windows_desktop_logins`.

<Admonition type="warning" title="RBAC Configuration">
Ensure that each Teleport user is only assigned Windows logins that they should
be allowed to access.
</Admonition>

Apply the new role to your cluster:

```code
$ tctl create -f windows-desktop-admins.yaml
```

(\!docs/pages/includes/add-role-to-user.mdx role="windows-desktop-admins" \!)

## Step 4/4. Connect

You can now connect to your Windows desktops from the Teleport Web UI:

> ![Connecting to a Windows desktop from the Web UI](../../img/desktop-access/non-ad-connect.png)

Reviewer: I recommend updating this screenshot, as the nav for Teleport 12 has changed, and it would be odd if the docs for a feature that is available in v12+ showed a UI from pre-v12.

Author: Updated

In this section we'll import the Teleport certificate authority (**CA**) file
to your Windows system, and prepare it for passwordless access through Teleport.

### Import the Teleport root certificate

Use `tctl` to export the Teleport user certificate authority:

```code
$ tctl auth export --type=windows > teleport.cer
```

Copy this certificate to your Windows system, if you didn't run `tctl` from there.

### Install the Teleport service for Windows

From the Windows system, download the [Teleport service installer](#LinkToCome) {/*Rename as needed */}
and run it. When prompted, select the Teleport certificate file from the previous step.
Once complete, reboot the system.

<Admonition type="tip" title="Headless Installation">
The Teleport service installer {/*Rename as needed*/} can be run in a shell
environment with elevated privileges. You can use this to automate installation,
uninstallation, and certificate updates. For example:

```code
$ teleport-windows-auth-setup-v12.0.0-amd64.exe install --cert=teleport.cer -r
Teleport Authentication Package installed
```

Use the `--help` flag to learn more.

</Admonition>

## Step 2/4. Install the Teleport Desktop Service

> On your local system authenticated to your Teleport cluster, generate a

Reviewer 2: ```suggestion
On your local system, authenticate to your Teleport cluster via `tsh` and generate a
```

Author: Isn't this already covered by `docs/pages/includes/tctl.mdx`, which has the user authenticate?

Reviewer 2: Yeah, that's true. I thought "On your local system authenticated to your Teleport cluster" could have been clearer. Maybe a comma after "system"?

short-lived join token:

```code
$ tctl tokens add --type=windowsdesktop
The invite token: (=presets.tokens.first=)
This token will expire in 60 minutes.

This token enables Desktop Access.  See https://goteleport.com/docs/desktop-access/
for detailed information on configuring Teleport Desktop Access with this token.
```

> Copy the token to the Linux host where you will run the Desktop service as `/tmp/token`.

Author: @ptgott please re-review this section. I realized while responding to other feedback that the instructions lacked a join token for the service node.

Reviewer 2: Added a minor suggestion
